sdk/how-to-guides/ports.md

```diff
@@ -0,0 +1,144 @@
+> See first: [Juju `open-port`](https://juju.is/docs/juju/hook-tool#open-port)
+
+**Contents**:
+
+1. [Open a port based on the charm config](#heading--open-a-port-to-your-charm)
+1. [Test the port is opened](#heading--test-the-port-is-opened)
+
+<a href="#heading--heading--open-a-port-to-your-charm"><h2 id="heading--heading--open-a-port-to-your-charm">Open a port based on the charm config</h2></a>
+
+To make a service permanently reachable under a stable URL on the cluster the charm needs to open (expose) a port. This ensures that the charm will be consistently accessible even if the pod gets recycled and the IP address changes.
+
+The port that the charm's service should be reachable on is typically a config option; for example, by including a section like this in `charmcraft.yaml`:
+
+```yaml
+options:
+  config:
+    server-port:
+      type: int
+      description: the port on which to offer the service
+      default: 8000
+```
+
+This means that the charm should update the open ports in the `config-changed` handler. In the `src/charm.py` file, in the `__init__` function of your charm, set up an observer for the `config-changed` event. For example:
+
+```python
+self.framework.observe(self.on.config_changed, self._on_config_changed)
+```
+
+> See more: `[`ops.ConfigChangedEvent`](https://ops.readthedocs.io/en/latest/#ops.ConfigChangedEvent)
+
+Now, in the body of the charm definition, define the event handler, and set the port that should be open - this defaults to a TCP port, but UDP and ICMP can also be specified. For example:
+
+```python
+def _on_config_changed(self, event: ops.ConfigChangedEvent):
+    self.unit.set_ports(self.config["server-port"])
+```
+
+> Examples: [`loki-k8s` sets the open port in the charm `__init__`](https://github.com/canonical/loki-k8s-operator/blob/2e4674d64b692f82ae3dec8f7f1f0745f087827f/src/charm.py#L150), [`mysql-router` sets ports based on whether the charm is configured to be externally accessible](https://github.com/canonical/mysql-router-operator/blob/04c7b3a0ebb6279648bd3f72c5c04b0f7bad1e0c/src/machine_charm.py#L111)
+
+> See more: [`ops.Unit.set_ports`](https://ops.readthedocs.io/en/latest/#ops.Unit.set_ports)
+
+<a href="#heading--test-the-port-is-opened"><h2 id="heading--test-the-port-is-opened">Test the port is opened</h2></a>
+
+> See first: [Get started with charm testing](https://juju.is/docs/sdk/get-started-with-charm-testing)
+
+You'll want to add three levels of tests, unit, Scenario, and integration.
```

Comment:
Punctuation and spelling suggestion: Add colon, uncapitalize "Scenario":

> You'll want to add three levels of tests: unit, scenario, and integration.

PS Now that scenarios is being promoted as the way to write unit tests, do we still need to distinguish?

Reply:
> uncapitalize "Scenario":

But "Scenario" is a proper noun, so it should have a capital, no? It's not a generic class of tests like "unit", "integration", "functional", and so on, it's tests that are written using the Scenario testing framework.

> PS Now that scenarios is being promoted as the way to write unit tests, do we still need to distinguish?

The PRs for actually getting Scenario into ops (when it will just be the ops.testing framework, and the "Scenario" name goes away) are in review at the moment. Once that's all finished, we should have a discussion about how we want to adjust the how-to guides (drop Harness immediately, make it less prominent, etc).